File: Knapsack/knapsack.cpp
// This C++ program solves the 0-1 knapsack problem. A very famous Dynamic Programming problem.
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

ll knapsack(ll V[], ll W[], ll n, ll maxWeight)
{
	ll K[n + 1][maxWeight + 1]; // Declaring the table for dynamic programming.
	for(ll i = 0; i <= n; i++)
	{
		for(ll j = 0; j <= maxWeight; j++)
		{
			if(i == 0 || j == 0)
				K[i][j] = 0; //Covering the base cases.

			else if(W[i - 1] <= j)// If the weight of the element is less than the capacity.
				K[i][j] = max(V[i - 1] + K[i - 1][j - W[i - 1]], K[i - 1][j]);// Then take the maximum of the value of this element or the leave this element.

			else
				K[i][j] = K[i - 1][j]; // If the weight of the element is more than the capacity, then leave this element.
		}
	}
	return K[n][maxWeight];
}

int main()
{
	ll n;
	cout << "Enter the value of n (The number of elements)." << endl;
	cin >> n;
	ll V[n], W[n];
	cout << "Enter " << n << " space separated integers denoting the values." << endl;
	for(ll i = 0; i < n; i++)
		cin >> V[i];
	cout << "Enter " << n << " space separated integers denoting the weights." << endl;
	for(ll i = 0; i < n; i++)
		cin >> W[i];
	ll maxWeight;
	cout << "Enter the value of the capacity of the knapsack" << endl;
	cin >> maxWeight;
	cout << "The maximum total value obtained with the following knapsack" << endl;
	cout << knapsack(V, W, n, maxWeight) << endl;
	return 0;
}

File: Competitive Coding/Dynamic Programming/0-1Knapsack/knapsack.java
import java.util.*;


class Knapsack
{
 
    // A function to return maximum of two integers
    static int maximum(int a, int b) 
      {
          if(a>b) 
          return a; 
          else 
          return b; 

      }
      
   // Returns the maximum value that can be put in a knapsack of capacity W
    static int  knapSack(int W, int wt[], int val[], int n)
    {
         int  i, w;
     int K[][] = new int[n+1][W+1];
      
     // Build table K[][] in bottom up manner
     for (i = 0; i <= n; i++)
     {
         for (w = 0; w <= W; w++)
         {
             if (i==0 || w==0)
                  K[i][w] = 0;
             else if (wt[i-1] <= w)
                   K[i][w] = maximum(val[i-1] + K[i-1][w-wt[i-1]] ,  K[i1][w]);
             else
                   K[i][w] = K[i-1][w];
         }
      }
      
      Return  K[n][W];
    }
 
   
   
    public static void main(String args[])
    {
		 
        Scanner s=new Scanner(System.in);
        System.out.println(“Enter number of objects”);
        int n=s.nextInt();
        int val[]=new int[n];
        int wt[] = new int[n];
        System.out.println(“Enter values of objects”);


        for(int i=0;i<n;i++)
          {
            val[i]=s.nextInt();
          }

        System.out.println(“Enter weights of object”);
        for(int i=0;i<n;i++)
          {
              wt[i]=s.nextInt();
          }


           System.out.println(“Enter the maximum weight”);
           int  W = s.nextInt();
           System.out.println(knapSack(W, wt, val, n));
            }
}

